File: rust/src/bin/day01.rs
// Day 1

use std::fs;

const INPUT_FILE: &str = "../inputs/day01.txt";

fn parse_input() -> Vec<i32> {
    fn parse(input: &str) -> i32 {
        let (direction, remainder) = input.split_at(1);
        let mut amount = remainder.parse::<i32>().expect("Cannot parse input");
        if direction == "L" {
            amount *= -1;
        }
        amount
    }

    fs::read_to_string(INPUT_FILE)
        .expect("No input file")
        .lines()
        .map(parse)
        .collect()
}

fn solve_p1(input: &[i32]) -> u32 {
    input
        .iter()
        .fold((50, 0_u32), |(pos, zeros), &rotation| {
            let pos = (pos + rotation).rem_euclid(100);
            let zeros = zeros + u32::from(pos == 0);
            (pos, zeros)
        })
        .1
}

fn solve_p2(input: &[i32]) -> u32 {
    let mut pos: i32 = 50;
    let mut zeros: u32 = 0;
    for &rotation in input {
        let start = pos;
        pos += rotation;
        zeros += pos.div_euclid(100).abs() as u32;
        if rotation.is_negative() && start == 0 {
            zeros -= 1;
        }
        pos = pos.rem_euclid(100);
        if pos == 0 && rotation.is_negative() {
            zeros += 1;
        }
    }
    zeros
}

fn main() {
    let input = parse_input();
    println!("Part 1: {}", solve_p1(&input));
    println!("Part 2: {}", solve_p2(&input));
}

File: python/src/day01.py
"""
Advent of Code 2025 day 1
"""

INPUT_FILE = "../inputs/day01.txt"


def parse_input():
    with open(INPUT_FILE, "r") as f:
        return [_.strip() for _ in f.readlines()]


def solve_p1(rotations):
    pos = 50
    zeros = 0
    for rot in rotations:
        d, x = rot[:1], int(rot[1:])
        if d == "L":
            x *= -1
        pos = (pos + x) % 100
        if pos == 0:
            zeros += 1
    return zeros


def solve_p2(rotations):
    pos = 50
    zeros = 0
    for rot in rotations:
        start = pos
        d, x = rot[:1], int(rot[1:])
        if d == "L":
            x *= -1
        pos += x
        passes, pos = divmod(pos, 100)
        if x < 0 and start == 0:
            # Already counted "passing" zero previously
            zeros += abs(passes) - 1
        else:
            zeros += abs(passes)
        if pos == 0 and x < 0:
            # Need to account for final land on zero from above
            zeros += 1
    return zeros


def main():
    data = parse_input()
    print(f"Part 1 = {solve_p1(data)}")
    print(f"Part 2 = {solve_p2(data)}")


if __name__ == "__main__":
    main()
